meta-balena-common/recipes-core/initrdscripts/files/fsuuidsinit

```diff
@@ -0,0 +1,102 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+#
+# Initializes unique filesystem UUIDs
+#
+
+# shellcheck disable=SC1091
+. /usr/libexec/os-helpers-logging
+. /usr/libexec/os-helpers-fs
+
+# Enable module function
+fsuuidsinit_enabled() {
+    # shellcheck disable=SC2154
+    if [ "$bootparam_flasher" = "true" ]; then
+        info "Flasher detected. Avoiding filesystem UUIDS initialization."
+        return 1
+    fi
+    # shellcheck disable=SC2154
+    default_uuid_prefix=$(get_cmdline_root_uuid | cut -d "-" -f1)
+    if [ "${default_uuid_prefix}" != "ba1eadef" ]; then
+        info "UUIDs already generated so no need to initialize any new ones."
+        return 1
+    fi
+    return 0
+}
+
+regenerate_uuid() {
+    dev=$1
+    # Filesystem type specific quirks
+    fstype=$2
+    old_uuid="$(get_dev_uuid "$dev")"
+    case $fstype in
+        vfat)
+            # mlabel doesn't work on file directly. It needs a drive to file
+            # mapping in its configuration.
+            drive=a:
+            info "Running quirks for $fstype..."
+            cat > "/etc/mtools.conf" <<EOF
+drive ${drive} file="$dev" exclusive
+EOF
+            cmd="mlabel -s -n"
+            if ! eval "${cmd}" "${drive}"; then
+                error "Failed to generate a new UUID for $dev."
+                return
+            fi
+            ;;
+        ext*)
+            # tune2fs refuses to work when the filesystem was not
+            # freshly checked.
+            info "Running quirks for $fstype..."
+            if ! e2fsck -fp "$dev"; then
+                warn "Failed to force check filesystem on $dev."
+            fi
+            cmd="echo y | EXT2FS_NO_MTAB_OK=1  tune2fs -U random"
+            if ! eval "${cmd}" "${dev}"; then
+                error "Failed to generate a new UUID for $dev."
+                return
+            fi
+            ;;
+        *) ;;
+        esac
+
+    new_uuid="$(get_dev_uuid "$dev")"
+    info "The new filesystem UUID for $dev is $new_uuid (from $old_uuid)."
+}
+
+# Main module function
+fsuuidsinit_run() {
+    # Generate new UUIDs/serials
+    info "Generating new UUIDs for all filesystems..."
+    for dev in $(blkid | cut -d ":" -f1); do
+        uuid="$(get_dev_uuid "${dev}")"
+        uuid_prefix="$(echo "${uuid}" | cut -d "-" -f1)"
+        fstype="$(get_dev_fstype "${dev}")"
+
+        if [ "${fstype}" = "vfat" ] && [ "${uuid}" != "BDEF-BDEF" ] ||
+           [ "${fstype}" = "ext4" ] && [ "${uuid_prefix}" != "ba1eadef"  ]; then
+               continue
```

Comment:
I believe the if should be like this instead:

> if (([ "${fstype}" = "vfat" ] && [ "${uuid}" != "BDEF-BDEF" ]) || ([ "${fstype}" = "ext4" ] && [ "${uuid_prefix}" != "ba1eadef"  ]))

Reply:
Parenthesis will spawn new subshells. I could group them with braces like:
```
        if { [ "${fstype}" = "vfat" ] && [ "${uuid}" != "BDEF-BDEF" ]; } ||
           { [ "${fstype}" = "ext4" ] && [ "${uuid_prefix}" != "ba1eadef"  ]; }; then
```
But in this case as bash precedence is from left to right, there is no difference in the result, is there?.

Reply:
We should use either (( or the {} grouping. Since (( spawns subshells, I guess we could avoid that and use { } grouping.
The problem I see with the way the PR has it is when: `fstype=vfat`, `uuid=\<some random uuid, different from BDEF-BDEF\>` and `uuid_prefix=ba1eadef`. This will make the code go through the else branch even though I believe that is no the intent.

Reply:
I see. I will use the braces as it is avoids problems.

The combination you mention cannot actually happen with the current implementation, as all the UUIDs are re-generated at the same time so the "ba1eadef" prefix can only appear with the default vfat BDEF-BDEF. But it's better to be safe, thanks.